IO_rot.f90

```diff
@@ -226,90 +231,167 @@ subroutine epilogue
+        call save_param_hdf5(Nexcite(:,:,:,3), "nexcite3.dac.", 3, (/INT(ix, KIND=8), INT(jx, KIND=8), INT(kx, KIND=8)/))
+        call save_param_hdf5(Nexcite(:,:,:,4), "nexcite4.dac.", 3, (/INT(ix, KIND=8), INT(jx, KIND=8), INT(kx, KIND=8)/))
+        call save_param_hdf5(Nexcite(:,:,:,5), "nexcite5.dac.", 3, (/INT(ix, KIND=8), INT(jx, KIND=8), INT(kx, KIND=8)/))
+        call save_param_hdf5(Nexcite(:,:,:,6), "nexcite6.dac.", 3, (/INT(ix, KIND=8), INT(jx, KIND=8), INT(kx, KIND=8)/))
+      endif
+      ! include viscosity results
+      if((flag_visc.ge.1).and.(vs_sav.eq.0)) then
+        call save1param(visc(:,:,:,1),tno//"viscx.dac.",1)
+        call save1param(visc(:,:,:,2),tno//"viscy.dac.",1)
+        call save1param(visc(:,:,:,3),tno//"viscz.dac.",1)
+        ! analogous save commands for HDF5 files
+        call save_param_hdf5(visc(:,:,:,1), "viscx.dac.", 3, (/INT(ix, KIND=8), INT(jx, KIND=8), INT(kx, KIND=8)/))
+        call save_param_hdf5(visc(:,:,:,2), "viscy.dac.", 3, (/INT(ix, KIND=8), INT(jx, KIND=8), INT(kx, KIND=8)/))
+        call save_param_hdf5(visc(:,:,:,3), "viscz.dac.", 3, (/INT(ix, KIND=8), INT(jx, KIND=8), INT(kx, KIND=8)/))
       endif
-       if((flag_visc.ge.1).and.(vs_sav.eq.0)) then
-          call save1param(visc(:,:,:,1),tno//"viscx.dac.",1)
-          call save1param(visc(:,:,:,2),tno//"viscy.dac.",1)
-          call save1param(visc(:,:,:,3),tno//"viscz.dac.",1)
-       endif
     endif
+
     if(flag_pip.eq.1 .or.flag_mhd.eq.0) then
-       do i=1,nvar_h
-          call save1param(U_h(:,:,:,i),tno//trim(file_h(i)),1)
-       enddo
+      do i=1,nvar_h
+        call save1param(U_h(:,:,:,i),tno//trim(file_h(i)),1)
+        call save_param_hdf5(U_h(:,:,:,i), trim(file_h(i)), 3, &
+          (/INT(ix, KIND=8), INT(jx, KIND=8), INT(kx, KIND=8)/))
+      enddo
     endif
-    if(flag_divb.eq.1 .and. flag_mhd.eq.1 .and. ps_sav .eq.0) then    
-       call save1param(U_m(:,:,:,9),tno//trim(file_m(9)),1)
+    ! Save divergence of B-field values (currently used in testing)
+    if(flag_divb.eq.1 .and. flag_mhd.eq.1 .and. ps_sav .eq.0) then
+      call save1param(U_m(:,:,:,9),tno//trim(file_m(9)),1)
+      call save_param_hdf5(U_m(:,:,:,9), trim(file_m(9)), 3, &
+        (/INT(ix, KIND=8), INT(jx, KIND=8), INT(kx, KIND=8)/))
     endif
-    
+
   end subroutine save_varfiles
 
+
   subroutine save1param(q,name,nvar)
     integer,intent(in)::nvar
     double precision, intent(in) :: q(ix,jx,kx,nvar)
     character(*), intent(in) :: name
     integer, parameter :: mf_q = 999
     integer nn
-    
+
     call dacdef3s(mf_q,trim(outdir) // '/' // name // cno,6,0)
     do nn=1,nvar
        write(mf_q) q(:,:,:,nn)
     enddo
 !print*,mf_q,name,q(1,1,1,:)
     close(mf_q)
   end subroutine save1param
-  
+
+  !
+  ! Saving a single variable array (of arbitrary rank) to an hdf5 file
+  !
+  subroutine save_param_hdf5(data, varname, rank, dims)
```

Comment:
The type conversion of the dummy arguments should take place here.

Reply:
Take place in the declaration of the subroutine? Given that `dims` is of an arbitrary size, I do not see how exactly this is done.

Reply:
I just pushed some changes that moves `dims` declaration inside `save_param_hdf5`. Makes the calls a lot cleaner.

Reply:
If I could figure out a way to get rank & dim-size from the `data` array inside the subroutine, `rank` and the explicit `dims` declarations would be unnecessary.